### PR TITLE
nixos/tor: fixes for settings.HidServAuth

### LINKED ISSUE #122429: Example for option services.tor.settings.HidServAuth
I've just upgraded from 20.09 to master, and found that `settings.tor.extraConfig` has been removed, in favour of `settings.tor.settings.*`, per RFC 42.

It's not clear to me how to use services.tor.settings.HidServAuth and each time I've tried, it fails.

### Attempt 1: Make it a list of attrs

```
services.tor.settings.HidServAuth = [
  {
    onion = "foo.onion";
    auth = "code";
  }
];
```

... fails with:

> error: value is a set while a list was expected, at /nix/store/aw790250d452a0m3psqhx91anj2rv6pq-source/lib/strings.nix:110:33

### Attempt 2: Make it an attrset.

```
services.tor.settings.HidServAuth.foo = {
  onion = "foo.onion";
  auth = "code";
};
```

... fails with:

> error: A definition for option `services.tor.settings.HidServAuth' is not of type `list of submodules'

This could be a bug, or I'm misusing. An example for `services.tor.settings.HidServAuth` could help either way. I'm not sure where in the docs to go about finding what a list of submodules should look like. I thought it was just "Attempt 1".

nixpkgs: a4b05310c2602e70b75b59c9189b8fd414c99bef

### PR DESCRIPTION
**Pull request by tomfitzhenry**
###### Motivation for this change
fixes https://github.com/NixOS/nixpkgs/issues/122429

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/security/tor.nix [hunk 1]
```diff
@@ -170,7 +170,7 @@ let
     else if k == "ServerTransportPlugin" then
       optionalString (v.transports != []) "${concatStringsSep "," v.transports} exec ${v.exec}"
     else if k == "HidServAuth" then
-      concatMapStringsSep "\n${k} " (settings: settings.onion + " " settings.auth) v
```
Comment (tomfitzhenry) [inline nixos/modules/services/security/tor.nix:173]: ju1m@, PTAL.

File: nixos/modules/services/security/tor.nix [hunk 2]
```diff
@@ -715,7 +715,7 @@ in
               (submodule {
                 options = {
                   onion = mkOption {
-                    type = strMatching "[a-z2-7]{16}(\\.onion)?";
+                    type = strMatching "[a-z2-7]{16,56}(\\.onion)?";
```
Comment (dotlambda) [inline nixos/modules/services/security/tor.nix:718]:
```suggestion
                    type = strMatching "([a-z2-7]{16}|[a-z2-7]{56})(\\.onion)?";
```

**Review by dotlambda — COMMENTED** [review]
Shouldn't we use `listsAsDuplicateKeys = true`?

**Comment by dotlambda** [comment]
Can we add a test case to `nixos/tests/tor.nix`?

File: nixos/modules/services/security/tor.nix (hunk 2 above)
Comment (dotlambda) [reply nixos/modules/services/security/tor.nix:718]: Btw, do you know if `.onion` is actually optional? That's not mentioned on https://2019.www.torproject.org/docs/tor-manual.html.en, but that documentation is outdated anyway and doesn't even mention v3 addresses.

**Review by ju1m — APPROVED** [review]

**Review by ju1m — CHANGES_REQUESTED** [review]
Support for v3 addresses in `HidServAuth` must not be added, besides LGTM.

**Comment by tomfitzhenry** [comment]
> Can we add a test case to nixos/tests/tor.nix?

I just tried this for 30 min, and it's surprisingly fiddly to test a hidden services. It requires setting up a directory authority, a relay, a server, a client, and coordinating various keys in various formats between the different actors.

**Comment by tomfitzhenry** [comment]
> @tomfitzhenry you may still want to use removeSuffix ".onion" v.onion + ".onion" or remove the ? in (\\.onion)?

Done. I opted for the latter, to avoid the maintenance burden of allowing different input types. This is consistent with torrc. Removing support for hostnames without ".onion" won't break anyone, since this functionality has never worked anyway.

**Review by ju1m — APPROVED** [review]

**Pull request merged**